File: Add Code Here/C++/tree/C++/Spiral_Matrix.cpp
#include<bits/stdc++.h>
using namespace std;
 vector<int> spiralOrder(vector<vector<int>>& matrix) {
    vector<int> ans;
    int n = matrix.size();
    int m = matrix[0].size();
        
    int top=0, bottom=n-1;
    int left=0, right=m-1;
        
    while(top<=bottom && left<=right)
    {
        for(int i=left; i<=right; i++)
        {
            ans.push_back(matrix[top][i]);
        }
        top++;
            
        for(int i=top; i<=bottom; i++)
        {
            ans.push_back(matrix[i][right]);
        }
        right--;
            
        if(top>bottom || left>right)
        {
            break;
        }
            
        for(int i=right; i>=left; i--)
        {
            ans.push_back(matrix[bottom][i]);
        }
        bottom--;
            
        for(int i=bottom; i>=top; i--)
        {
            ans.push_back(matrix[i][left]);
        }
        left++;
    }
        
    return ans;
}

int main()
{
	int n,m;
	cin>>n>>m;

	vector<vector<int>> matrix(n, vector<int>(m));
	for(int i=0; i<n; i++)
	{
		for(int j=0; j<m; j++)
		{
			cin>>matrix[i][j];
		}
	}
	vector<int> ans = spiralOrder(matrix);
	for(int i=0; i<ans.size(); i++)
	{
		cout<<ans[i]<<" ";
	}
	cout<<endl;
}

File: Hacktoberfest2023-main/Add Code Here/JAVA/SpiralMatrix.java
public class SpiralMatrix {
    public static void main(String[] args) {
        int n = 4; // Change this value to the desired size of the matrix
        int[][] spiralMatrix = generateSpiralMatrix(n);

        // Print the spiral matrix
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                System.out.printf("%2d ", spiralMatrix[i][j]);
            }
            System.out.println();
        }
    }

    public static int[][] generateSpiralMatrix(int n) {
        int[][] matrix = new int[n][n];
        int num = 1;
        int left = 0, right = n - 1, top = 0, bottom = n - 1;

        while (left <= right && top <= bottom) {
            // Traverse from left to right
            for (int i = left; i <= right; i++) {
                matrix[top][i] = num++;
            }
            top++;

            // Traverse from top to bottom
            for (int i = top; i <= bottom; i++) {
                matrix[i][right] = num++;
            }
            right--;

            // Traverse from right to left
            for (int i = right; i >= left; i--) {
                matrix[bottom][i] = num++;
            }
            bottom--;

            // Traverse from bottom to top
            for (int i = bottom; i >= top; i--) {
                matrix[i][left] = num++;
            }
            left++;
        }

        return matrix;
    }
}
